cpp/src/slice2cs/Gen.cpp

```diff
@@ -2025,49 +2004,99 @@ Slice::Gen::TypesVisitor::visitClassDefEnd(const ClassDefPtr& p)
     {
         _out << sp << nl << "#region Constructors";
 
-        _out << sp;
-        emitGeneratedCodeAttribute();
-        _out << nl << "public " << name << spar << epar;
-        if (base)
-        {
-            _out << " : base()";
-        }
-        _out << sb;
-        writeDataMemberInitializers(dataMembers, DotNet::ICloneable);
-        _out << nl << "ice_initialize();";
-        _out << eb;
-
+        // Primary constructor.
         _out << sp;
         emitGeneratedCodeAttribute();
         _out << nl << "public " << name << spar;
+
         vector<string> paramDecl;
-        for (DataMemberList::const_iterator d = allDataMembers.begin(); d != allDataMembers.end(); ++d)
+        vector<string> secondaryCtorParams;
+        vector<string> secondaryCtorMemberNames;
+        vector<string> secondaryCtorBaseParamNames;
+        for (const auto& q : allDataMembers)
         {
-            string memberName = fixId((*d)->name(), DotNet::ICloneable);
-            string memberType = typeToString((*d)->type(), ns, (*d)->optional());
+            string memberName = fixId(q->name(), DotNet::ICloneable);
+            string memberType = typeToString(q->type(), ns, q->optional());
             paramDecl.push_back(memberType + " " + memberName);
+
+            // Look for non-nullable fields to be initialized by the secondary constructor.
+            if (!q->optional())
+            {
+                StructPtr st = dynamic_pointer_cast<Struct>(q->type());
+                if (dynamic_pointer_cast<Sequence>(q->type()) || dynamic_pointer_cast<Dictionary>(q->type()) ||
+                    (st && isMappedToClass(st)))
+                {
+                    secondaryCtorParams.push_back(memberType + " " + memberName);
+
+                    if (find(dataMembers.begin(), dataMembers.end(), q) != dataMembers.end())
+                    {
+                        secondaryCtorMemberNames.push_back(memberName);
+                    }
+                }
+            }
         }
         _out << paramDecl << epar;
+
         if (base && allDataMembers.size() != dataMembers.size())
         {
             _out << " : base" << spar;
             vector<string> baseParamNames;
             DataMemberList baseDataMembers = base->allDataMembers();
-            for (DataMemberList::const_iterator d = baseDataMembers.begin(); d != baseDataMembers.end(); ++d)
+            for (const auto& q : baseDataMembers)
             {
-                baseParamNames.push_back(fixId((*d)->name(), DotNet::ICloneable));
+                string memberName = fixId(q->name(), DotNet::ICloneable);
+                baseParamNames.push_back(memberName);
+
```

Comment:
Shouldn't we also check !optional for `secondaryCtorBaseParamNames`?

Reply:
You're right, I'll fix it in my next PR.